include/boost/numeric/ublas/tensor/detail/extents_functions.hpp

```diff
@@ -0,0 +1,366 @@
+        // to comply with GNU Octave this check is made
+        // if position 2 contains 1 we push at back
+        // else we push at front
+        if constexpr( ExtentsType::at(1) == 1ul ){
+          return decltype( number_list_to_static_extents<value_type>(
+                    decltype( push_back<value_type(1)>(one_free_num_list)){} ) 
+                  ){};
+        }else{
+          return decltype( number_list_to_static_extents<value_type>(
+                    decltype( push_front<value_type(1)>(one_free_num_list)){} ) 
+                  ){};
+        }
+
+      }else{
+        return decltype( number_list_to_static_extents<value_type>(one_free_num_list) ){};
+
+      }
+
+    }else{
+      return decltype( number_list_to_static_extents<value_type>( num_list ) ) {};
+    }
+    
+  }
+
+  template <class ExtentsType, typename std::enable_if<is_dynamic<ExtentsType>::value && is_dynamic_rank<ExtentsType>::value, int>::type = 0>
+  BOOST_UBLAS_INLINE
+  constexpr auto squeeze_impl( ExtentsType const& e ){
+    
+    if( e.size() <= 2 ){
+      return e;
+    }
+
+    using base_type = typename ExtentsType::base_type;
+    using value_type = typename ExtentsType::value_type;
+    base_type n_extents;
+
+    // copying non 1s to the new extents
+    std::copy_if(e.begin(), e.end(), std::back_inserter(n_extents), [](auto const& el){
+      return el != value_type(1);
+    });
+    
+    // checking if extents size goes blow 2
+    switch( n_extents.size() ){
+      // if size of extents goes to zero
+      // return (1,1)
+      case 0ul:{
+        return ExtentsType{value_type(1),value_type(1)};
+      }
+      // if size of extents goes to 1
+      // complying with GNU Octave
+      // if position 2 contains 1 we push at back
+      // else we push at front
+      case 1ul:{
+        if( e[1] != value_type(1) ){
+          n_extents.insert(n_extents.begin(), value_type(1));
+        }else{
+          n_extents.push_back(value_type(1));
+        }
+        [[fallthrough]];
+      }
+      default:{
+        return ExtentsType(n_extents); 
+      }
+    }
+    
+  }
+
+  template <class ExtentsType, typename std::enable_if<is_dynamic<ExtentsType>::value && is_static_rank<ExtentsType>::value, int>::type = 0>
+  BOOST_UBLAS_INLINE
+  constexpr auto squeeze_impl( ExtentsType const& e ){
+    if constexpr( ExtentsType::size() <= 2 ){
+      return e;
+    }else{
+      return squeeze_impl(e.to_dynamic_extents());
+    }
+  }
+    
+    
+
+} // namespace boost::numeric::ublas::detail
+
+
+namespace boost::numeric::ublas {
+
+/** @brief Returns true if size > 1 and all elements > 0 or size == 1 && e[0] == 1 */
+template <class ExtentsType, typename std::enable_if<detail::is_extents<ExtentsType>::value, int>::type = 0>
+BOOST_UBLAS_INLINE 
+constexpr bool valid(ExtentsType const &e) {
+  if ( e.size() == 1 && e[0] == 1 ){
+    return true;
+  }
+  return ( e.size() > typename ExtentsType::size_type(1) ) &&
```

Comment:
I think extents are non-neg values. But if you want I will change it all_of.

Reply:
Correct. However, the template parameter `ExtentsType` are not restricted to only positive values.

Reply:
we are static asserting about the sign. So no one can create neg valued extents.

Reply:
why don't we use comments for telling the condition because compilers will give warning that no return type warning and if someone has enabled warning as the error it will not compile.

Reply:
Or we could define macro for if-else where else will have condition.

Reply:
I am okay with `[](auto a) { return a != 0u; }`. 
Do not like macro.
I think `[](auto a) { return a > 0u; }` will not hurt.